### PR TITLE
feat: add nft payment guarantee module

### LINKED ISSUE #5: Create a contract to trade NFTs.
We need to implement the contract for NFT transactions.
Consider the following points:

* You must verify that the NFT is from a valid contract.
* We need to know the price of NFT. We currently utilize NFT's Description.
* You must be able to manage the NFT you want to trade in this contract. The owner of the NFT must be verified to be a contract.
* If the terms and conditions of the contract are met, the owner of the NFT must be the payer, and the existing NFT owner must be paid the money(MED).
* If the contract does not match, the owner of the NFT must be changed to the previous owner.
* The payer must be able to collect the amount.

## Functions
- [x] Deposit in contract
- [x] NFT transactions
- [ ] Refund deposited amount(Only payer who created the payment contract)
- [ ] Change owner of NFT to previous owner
- [ ] Change owner of NFT to previous owner in case of transaction failure
- [ ] Add/Delete valid NFT contracts

## Validation
- [x] The contract address of the valid NFT shall be specified in the contract for the transaction.
- [x] The price of the NFT shall be deposited in the payment contract.
- [x] The owner of the NFT shall be the payment contract.(Payment contract must be able to change ownership of NFT)

### PR DESCRIPTION
**Pull request by gyuguen**
### Feature
* It will receive nft and pay the corresponding amount.
* The account that generated this contract will own nft.
* Check the contract in which nft is generated, and verify that the owner of nft is this contract.
* (To be implemented) If there is not enough deposit in the contract, nft will return to its original owner.
* There is a sufficient deposit in the contract, and upon completion of NFT verification, the owner of NFT will be the creator of this contract, and the amount specified in NFT will be paid to the previous NFT owner.

I will be update to READ.ME

### CODE REVIEW

File: payment-guarantee/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let contract_info = ContractInfo {
+        source_contracts: msg.source_contracts,
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs:20]: Value added to determine that only NFT sent by a specific contract is valid.
I think we also need to modify this value later.

File: payment-guarantee/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        ExecuteMsg::Deposit {} => execute_deposit(deps, env, info),
+        ExecuteMsg::ReceiveNft(msg) => execute_receive_nft(deps, env, info, msg),
+    }
+}
+
+pub fn execute_deposit(
+    mut _deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs:46]: It's just putting in a deposit

File: payment-guarantee/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+pub fn execute_receive_nft(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let source_contract = info.sender.clone();
+    let contract_info = CONTRACT_INFO.load(deps.storage)?;
+    if is_invalid_from_contract(&contract_info, source_contract.to_string()) {
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs:70]: Verify that it is an NFT from a valid contract.

File: payment-guarantee/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    let token_id = msg.token_id.to_string();
+    let query_msg = QueryMsg::OwnerOf {
+        token_id: token_id.to_string(),
+        include_expired: None,
+    };
+
+    let res: OwnerOfResponse = deps.querier.query_wasm_smart(source_contract.as_str(), &query_msg)?;
+    if env.contract.address.to_string().ne(res.owner.as_str()) {
+        return Err(ContractError::Unauthorized { msg: "The owner of the token must be this contract.".to_string() });
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs]: Verify that the owner of the NFT is this contract.

File: payment-guarantee/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    let res: StdResult<TokenInfoMsg> = from_binary(&msg.msg.unwrap());
+    if res.is_err() {
+        return Err(ContractError::InvalidParameter { msg: "msg is invalid parameter.".to_string() });
+    }
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs]: Decode to determine the price of the NFT.
Invalid NFT if decoding is not possible.

File: payment-guarantee/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+    let token_info_msg = res.unwrap();
+    let token_price = token_info_msg.price;
+
+    let deposit_coin = deps.querier.query_balance(env.contract.address, token_price.denom.clone())?;
+    if deposit_coin.amount.le(token_price.amount.borrow()) {
+        return Err(ContractError::InsufficientDeposit {});
+    }
```
Comment (gyuguen) [inline payment-guarantee/src/contract.rs]: If the current Contract does not have enough deposit, the transaction will fail.
The ability to change the owner of NFT to an existing owner should be added later.

File: payment-guarantee/payment-guarantee.iml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  <component name="NewModuleRootManager" inherit-compiler-output="true">
+    <exclude-output />
+    <content url="file://$MODULE_DIR$">
+      <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
+      <excludeFolder url="file://$MODULE_DIR$/target" />
+    </content>
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Comment (youngjoon-lee) [inline payment-guarantee/payment-guarantee.iml]: I guess we can remove this file.

File: payment-guarantee/src/contract.rs (hunk 1 above)
Comment (youngjoon-lee) [reply payment-guarantee/src/contract.rs:20]: Does it mean that the `source_contracts` is for determining that only NFT sent by a specific contract is valid?

Comment (gyuguen) [reply payment-guarantee/src/contract.rs:20]: Yes. You're right.

File: payment-guarantee/payment-guarantee.iml (hunk 7 above)
Comment (gyuguen) [reply payment-guarantee/payment-guarantee.iml]: That's right. I'll remove it.

File: payment-guarantee/src/contract.rs [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let source_contract = info.sender.clone();
+    let contract_info = CONTRACT_INFO.load(deps.storage)?;
+    if is_invalid_from_contract(&contract_info, source_contract.to_string()) {
+        return Err(ContractError::Unauthorized {
+            msg: format!("The token has a different contract. source: {}, this: {}", source_contract.as_str(), contract_info.source_contracts.join(",")),
```
Comment (youngjoon-lee) [inline payment-guarantee/src/contract.rs]:
```suggestion
            msg: format!("The token belongs to an unexpected contract. actual: {}, expected: {}", source_contract.as_str(), contract_info.source_contracts.join(",")),
```

File: payment-guarantee/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    };
+
+    let res: OwnerOfResponse = deps.querier.query_wasm_smart(source_contract.as_str(), &query_msg)?;
+    if env.contract.address.to_string().ne(res.owner.as_str()) {
+        return Err(ContractError::Unauthorized { msg: "The owner of the token must be this contract.".to_string() });
+    }
+
+    let res: StdResult<TokenInfoMsg> = from_binary(&msg.msg.unwrap());
+    if res.is_err() {
+        return Err(ContractError::InvalidParameter { msg: "msg is invalid parameter.".to_string() });
```
Comment (youngjoon-lee) [inline payment-guarantee/src/contract.rs]:
```suggestion
        return Err(ContractError::InvalidParameter { msg: "msg is invalid".to_string() });
```
or `msg is an invalid parameter`

File: payment-guarantee/src/msg.rs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct TokenInfoMsg {
+    pub contract: String,
+    pub description: Option<String>,
+    pub price: Coin,
+    pub sender: String,
+}
```
Comment (youngjoon-lee) [inline payment-guarantee/src/msg.rs:26]: It seems that the `contract` and the `description` aren't used anywhere.

Comment (gyuguen) [reply payment-guarantee/src/msg.rs:26]: That's right. I'm not using it right now.
The current development is minimal verification and exchange.
`contract` will be used for verification.
Description will be shown as described in NFT when showing contract status.(with token_id, name, etc...)

Comment (gyuguen) [reply payment-guarantee/src/msg.rs:26]: In addition, Successful NFT transactions are achieved on NFT transmission (validated and exchanged immediately).

I'm thinking about what action to take to exchange the NFT.
* When NFT was sent (currently implemented)
* Alternatively, perform verification and exchange operations in a separate action.

Comment (youngjoon-lee) [reply payment-guarantee/src/msg.rs:26]: Is the `TokenInfoMsg` only for the `Cw721ReceiveMsg.msg`? Then, only the `sender` would be enough.

I'm confused what's the difference with `TokenInfoMsg` and `TokenInfo` (in the PR #3). I saw that the `TokenInfo` is put into the `cw721-base.TokenInfo.description` as JSON. That's good.
Then, I guess we can unmarshal the `cw721-base.TokenInfo.description` when we handle the `Cw721ReceiveMsg`.
Then, we can get `contract`, `description` and `price` from it.

I'm not sure if it's possible, but it looks much simpler because we don't need any duplicated declarations.

Comment (gyuguen) [reply payment-guarantee/src/msg.rs:26]: TokenInfo, located in `panacea-nft`, is used to store additional information when MINT the NFT.
As a result, if only a `sender` is added, it is the same structure as TokenInfoMsg.
The `sender` is only required to perform SEND_NFT, so it is not included in the Structures field used to save.

Since panacea-nft and payment-guarantee are different projects, declarations on structures have been made respectively.
- In `panacea-nft`, declare to `TokenInfo` and `TokenInfoWithSender`
- In `payment-guarantee`, declare to `TokenInfoMsg`.

I think `TokenInfoMsg` should be used to declare a common structure. 
However, we can distribute it to [crates.io](https://crates.io/) and manage it or declare it as below.
`payment-guarantee = { path = "../payment-guarantee", version = "0.0.1" }`

Comment (youngjoon-lee) [reply payment-guarantee/src/msg.rs:26]: Let me rephrase the topic.

In my guess,
In `payment-guarantee`, the `TokenInfoMsg` can be defined as below:
```
pub struct TokenInfoMsg {
    pub sender: String,
}
```
Also, we can rename `TokenInfoMsg` to `ReceiveNftMsgBody` or something else, because that isn't the token info. That's the additional info for handling the `ReceiveNft` operation.

If you need the token info when handling the `ReceiveNft` operation, I think we can get the `TokenInfo` (defined in the `panacea-nft`) using `deps.querier.query_wasm_smart()`.
I didn't try it yet, but it seems it's feasible.

Then, as you said, we should do:
```
panacea-nft = { path = "../panacea-nft", version = "0.0.1" }
```
(IMHO, the `panacea-nft` is the mother project of the `payment-guarantee`, not the other way around.)

Comment (youngjoon-lee) [reply payment-guarantee/src/msg.rs:26]: My main concern is that it looks inflexible to declare duplicated fields (`contract`, `description` and `price`) in both of `TokenInfo` and `TokenInfoMsg`.

Comment (youngjoon-lee) [reply payment-guarantee/src/msg.rs:26]: And, please put `TODO` comments about what further validations should be added in the future.

Comment (gyuguen) [reply payment-guarantee/src/msg.rs:26]: Aha, I understand. 
Likewise, I also thought about duplicate fields.
The method you suggested is an implementable method.

When inquiring in `deps.querier` to verify that the NFT is contract-owned, I think it would be better to other information be imported and used together.(set price)

**Review by youngjoon-lee — APPROVED** [review]
Awesome!

File: payment-guarantee/src/msg.rs (hunk 10 above)
Comment (gyuguen) [reply payment-guarantee/src/msg.rs:26]: > And, please put `TODO` comments about what further validations should be added in the future.

I'm writing what needs to be developed and validation here.

https://github.com/medibloc/panacea-smart-contract/issues/5
